File: src/ObjectManager/Object.Tes/FilePacks/Records/045-EFSH.Effect Shader.cs
using OA.Core;

namespace OA.Tes.FilePacks.Records
{
    public class EFSHRecord : Record
    {
        public class DATAField
        {
            public byte Flags;
            public uint MembraneShader_SourceBlendMode;
            public uint MembraneShader_BlendOperation;
            public uint MembraneShader_ZTestFunction;
            public ColorRef FillTextureEffect_Color;
            public float FillTextureEffect_AlphaFadeInTime;
            public float FillTextureEffect_FullAlphaTime;
            public float FillTextureEffect_AlphaFadeOutTime;
            public float FillTextureEffect_PresistentAlphaRatio;
            public float FillTextureEffect_AlphaPulseAmplitude;
            public float FillTextureEffect_AlphaPulseFrequency;
            public float FillTextureEffect_TextureAnimationSpeed_U;
            public float FillTextureEffect_TextureAnimationSpeed_V;
            public float EdgeEffect_FallOff;
            public ColorRef EdgeEffect_Color;
            public float EdgeEffect_AlphaFadeInTime;
            public float EdgeEffect_FullAlphaTime;
            public float EdgeEffect_AlphaFadeOutTime;
            public float EdgeEffect_PresistentAlphaRatio;
            public float EdgeEffect_AlphaPulseAmplitude;
            public float EdgeEffect_AlphaPulseFrequency;
            public float FillTextureEffect_FullAlphaRatio;
            public float EdgeEffect_FullAlphaRatio;
            public uint MembraneShader_DestBlendMode;
            public uint ParticleShader_SourceBlendMode;
            public uint ParticleShader_BlendOperation;
            public uint ParticleShader_ZTestFunction;
            public uint ParticleShader_DestBlendMode;
            public float ParticleShader_ParticleBirthRampUpTime;
            public float ParticleShader_FullParticleBirthTime;
            public float ParticleShader_ParticleBirthRampDownTime;
            public float ParticleShader_FullParticleBirthRatio;
            public float ParticleShader_PersistantParticleBirthRatio;
            public float ParticleShader_ParticleLifetime;
            public float ParticleShader_ParticleLifetime_Delta;
            public float ParticleShader_InitialSpeedAlongNormal;
            public float ParticleShader_AccelerationAlongNormal;
            public float ParticleShader_InitialVelocity1;
            public float ParticleShader_InitialVelocity2;
            public float ParticleShader_InitialVelocity3;
            public float ParticleShader_Acceleration1;
            public float ParticleShader_Acceleration2;
            public float ParticleShader_Acceleration3;
            public float ParticleShader_ScaleKey1;
            public float ParticleShader_ScaleKey2;
            public float ParticleShader_ScaleKey1Time;
            public float ParticleShader_ScaleKey2Time;
            public ColorRef ColorKey1_Color;
            public ColorRef ColorKey2_Color;
            public ColorRef ColorKey3_Color;
            public float ColorKey1_ColorAlpha;
            public float ColorKey2_ColorAlpha;
            public float ColorKey3_ColorAlpha;
            public float ColorKey1_ColorKeyTime;
            public float ColorKey2_ColorKeyTime;
            public float ColorKey3_ColorKeyTime;

            public DATAField(UnityBinaryReader r, int dataSize)
            {
                if (dataSize != 224 && dataSize != 96)
                    Flags = 0;
                Flags = r.ReadByte();
                r.SkipBytes(3); // Unused
                MembraneShader_SourceBlendMode = r.ReadLEUInt32();
                MembraneShader_BlendOperation = r.ReadLEUInt32();
                MembraneShader_ZTestFunction = r.ReadLEUInt32();
                FillTextureEffect_Color = new ColorRef(r);
                FillTextureEffect_AlphaFadeInTime = r.ReadLESingle();
                FillTextureEffect_FullAlphaTime = r.ReadLESingle();
                FillTextureEffect_AlphaFadeOutTime = r.ReadLESingle();
                FillTextureEffect_PresistentAlphaRatio = r.ReadLESingle();
                FillTextureEffect_AlphaPulseAmplitude = r.ReadLESingle();
                FillTextureEffect_AlphaPulseFrequency = r.ReadLESingle();
                FillTextureEffect_TextureAnimationSpeed_U = r.ReadLESingle();
                FillTextureEffect_TextureAnimationSpeed_V = r.ReadLESingle();
                EdgeEffect_FallOff = r.ReadLESingle();
                EdgeEffect_Color = new ColorRef(r);
                EdgeEffect_AlphaFadeInTime = r.ReadLESingle();
                EdgeEffect_FullAlphaTime = r.ReadLESingle();
                EdgeEffect_AlphaFadeOutTime = r.ReadLESingle();
                EdgeEffect_PresistentAlphaRatio = r.ReadLESingle();
                EdgeEffect_AlphaPulseAmplitude = r.ReadLESingle();
                EdgeEffect_AlphaPulseFrequency = r.ReadLESingle();
                FillTextureEffect_FullAlphaRatio = r.ReadLESingle();
                EdgeEffect_FullAlphaRatio = r.ReadLESingle();
                MembraneShader_DestBlendMode = r.ReadLEUInt32();
                if (dataSize == 96)
                    return;
                ParticleShader_SourceBlendMode = r.ReadLEUInt32();
                ParticleShader_BlendOperation = r.ReadLEUInt32();
                ParticleShader_ZTestFunction = r.ReadLEUInt32();
                ParticleShader_DestBlendMode = r.ReadLEUInt32();
                ParticleShader_ParticleBirthRampUpTime = r.ReadLESingle();
                ParticleShader_FullParticleBirthTime = r.ReadLESingle();
                ParticleShader_ParticleBirthRampDownTime = r.ReadLESingle();
                ParticleShader_FullParticleBirthRatio = r.ReadLESingle();
                ParticleShader_PersistantParticleBirthRatio = r.ReadLESingle();
                ParticleShader_ParticleLifetime = r.ReadLESingle();
                ParticleShader_ParticleLifetime_Delta = r.ReadLESingle();
                ParticleShader_InitialSpeedAlongNormal = r.ReadLESingle();
                ParticleShader_AccelerationAlongNormal = r.ReadLESingle();
                ParticleShader_InitialVelocity1 = r.ReadLESingle();
                ParticleShader_InitialVelocity2 = r.ReadLESingle();
                ParticleShader_InitialVelocity3 = r.ReadLESingle();
                ParticleShader_Acceleration1 = r.ReadLESingle();
                ParticleShader_Acceleration2 = r.ReadLESingle();
                ParticleShader_Acceleration3 = r.ReadLESingle();
                ParticleShader_ScaleKey1 = r.ReadLESingle();
                ParticleShader_ScaleKey2 = r.ReadLESingle();
                ParticleShader_ScaleKey1Time = r.ReadLESingle();
                ParticleShader_ScaleKey2Time = r.ReadLESingle();
                ColorKey1_Color = new ColorRef(r);
                ColorKey2_Color = new ColorRef(r);
                ColorKey3_Color = new ColorRef(r);
                ColorKey1_ColorAlpha = r.ReadLESingle();
                ColorKey2_ColorAlpha = r.ReadLESingle();
                ColorKey3_ColorAlpha = r.ReadLESingle();
                ColorKey1_ColorKeyTime = r.ReadLESingle();
                ColorKey2_ColorKeyTime = r.ReadLESingle();
                ColorKey3_ColorKeyTime = r.ReadLESingle();
            }
        }

        public override string ToString() => $"EFSH: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID
        public FILEField ICON; // Fill Texture
        public FILEField ICO2; // Particle Shader Texture
        public DATAField DATA; // Data

        public override bool CreateField(UnityBinaryReader r, GameFormatId format, string type, int dataSize)
        {
            switch (type)
            {
                case "EDID": EDID = new STRVField(r, dataSize); return true;
                case "ICON": ICON = new FILEField(r, dataSize); return true;
                case "ICO2": ICO2 = new FILEField(r, dataSize); return true;
                case "DATA": DATA = new DATAField(r, dataSize); return true;
                default: return false;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/045-EFSH.Effect Shader.swift
//
//  EFSHRecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class EFSHRecord: Record {
    public struct DATAField {
        public let flags: UInt8
        public let membraneShader_sourceBlendMode: UInt32
        public let membraneShader_blendOperation: UInt32
        public let membraneShader_ztestFunction: UInt32
        public let fillTextureEffect_color: ColorRef4
        public let fillTextureEffect_alphaFadeInTime: Float
        public let fillTextureEffect_fullAlphaTime: Float
        public let fillTextureEffect_alphaFadeOutTime: Float
        public let fillTextureEffect_presistentAlphaRatio: Float
        public let fillTextureEffect_alphaPulseAmplitude: Float
        public let fillTextureEffect_alphaPulseFrequency: Float
        public let fillTextureEffect_textureAnimationSpeed_u: Float
        public let fillTextureEffect_textureAnimationSpeed_v: Float
        public let edgeEffect_fallOff: Float
        public let edgeEffect_color: ColorRef4
        public let edgeEffect_alphaFadeInTime: Float
        public let edgeEffect_fullAlphaTime: Float
        public let edgeEffect_alphaFadeOutTime: Float
        public let edgeEffect_presistentAlphaRatio: Float
        public let edgeEffect_alphaPulseAmplitude: Float
        public let edgeEffect_alphaPulseFrequency: Float
        public let fillTextureEffect_fullAlphaRatio: Float
        public let edgeEffect_fullAlphaRatio: Float
        public let membraneShader_destBlendMode: UInt32
        public var particleShader_sourceBlendMode: UInt32? = nil
        public var particleShader_blendOperation: UInt32? = nil
        public var particleShader_ztestFunction: UInt32? = nil
        public var particleShader_destBlendMode: UInt32? = nil
        public var particleShader_particleBirthRampUpTime: Float? = nil
        public var particleShader_fullParticleBirthTime: Float? = nil
        public var particleShader_particleBirthRampDownTime: Float? = nil
        public var particleShader_fullParticleBirthRatio: Float? = nil
        public var particleShader_persistantParticleBirthRatio: Float? = nil
        public var particleShader_particleLifetime: Float? = nil
        public var particleShader_particleLifetime_Delta: Float? = nil
        public var particleShader_initialSpeedAlongNormal: Float? = nil
        public var particleShader_accelerationAlongNormal: Float? = nil
        public var particleShader_initialVelocity1: Float? = nil
        public var particleShader_initialVelocity2: Float? = nil
        public var particleShader_initialVelocity3: Float? = nil
        public var particleShader_acceleration1: Float? = nil
        public var particleShader_acceleration2: Float? = nil
        public var particleShader_acceleration3: Float? = nil
        public var particleShader_scaleKey1: Float? = nil
        public var particleShader_scaleKey2: Float? = nil
        public var particleShader_scaleKey1Time: Float? = nil
        public var particleShader_scaleKey2Time: Float? = nil
        public var colorKey1_color: ColorRef4? = nil
        public var colorKey2_color: ColorRef4? = nil
        public var colorKey3_color: ColorRef4? = nil
        public var colorKey1_colorAlpha: Float? = nil
        public var colorKey2_colorAlpha: Float? = nil
        public var colorKey3_colorAlpha: Float? = nil
        public var colorKey1_colorKeyTime: Float? = nil
        public var colorKey2_colorKeyTime: Float? = nil
        public var colorKey3_colorKeyTime: Float? = nil

        init(_ r: BinaryReader, _ dataSize: Int) {
            flags = r.readByte()
            r.skipBytes(3) // Unused
            membraneShader_sourceBlendMode = r.readLEUInt32()
            membraneShader_blendOperation = r.readLEUInt32()
            membraneShader_ztestFunction = r.readLEUInt32()
            fillTextureEffect_color = r.readT(4)
            fillTextureEffect_alphaFadeInTime = r.readLESingle()
            fillTextureEffect_fullAlphaTime = r.readLESingle()
            fillTextureEffect_alphaFadeOutTime = r.readLESingle()
            fillTextureEffect_presistentAlphaRatio = r.readLESingle()
            fillTextureEffect_alphaPulseAmplitude = r.readLESingle()
            fillTextureEffect_alphaPulseFrequency = r.readLESingle()
            fillTextureEffect_textureAnimationSpeed_u = r.readLESingle()
            fillTextureEffect_textureAnimationSpeed_v = r.readLESingle()
            edgeEffect_fallOff = r.readLESingle()
            edgeEffect_color = r.readT(4)
            edgeEffect_alphaFadeInTime = r.readLESingle()
            edgeEffect_fullAlphaTime = r.readLESingle()
            edgeEffect_alphaFadeOutTime = r.readLESingle()
            edgeEffect_presistentAlphaRatio = r.readLESingle()
            edgeEffect_alphaPulseAmplitude = r.readLESingle()
            edgeEffect_alphaPulseFrequency = r.readLESingle()
            fillTextureEffect_fullAlphaRatio = r.readLESingle()
            edgeEffect_fullAlphaRatio = r.readLESingle()
            membraneShader_destBlendMode = r.readLEUInt32()
            guard dataSize != 96 else {
                return
            }
            particleShader_sourceBlendMode = r.readLEUInt32()
            particleShader_blendOperation = r.readLEUInt32()
            particleShader_ztestFunction = r.readLEUInt32()
            particleShader_destBlendMode = r.readLEUInt32()
            particleShader_particleBirthRampUpTime = r.readLESingle()
            particleShader_fullParticleBirthTime = r.readLESingle()
            particleShader_particleBirthRampDownTime = r.readLESingle()
            particleShader_fullParticleBirthRatio = r.readLESingle()
            particleShader_persistantParticleBirthRatio = r.readLESingle()
            particleShader_particleLifetime = r.readLESingle()
            particleShader_particleLifetime_Delta = r.readLESingle()
            particleShader_initialSpeedAlongNormal = r.readLESingle()
            particleShader_accelerationAlongNormal = r.readLESingle()
            particleShader_initialVelocity1 = r.readLESingle()
            particleShader_initialVelocity2 = r.readLESingle()
            particleShader_initialVelocity3 = r.readLESingle()
            particleShader_acceleration1 = r.readLESingle()
            particleShader_acceleration2 = r.readLESingle()
            particleShader_acceleration3 = r.readLESingle()
            particleShader_scaleKey1 = r.readLESingle()
            particleShader_scaleKey2 = r.readLESingle()
            particleShader_scaleKey1Time = r.readLESingle()
            particleShader_scaleKey2Time = r.readLESingle()
            colorKey1_color = r.readT(4)
            colorKey2_color = r.readT(4)
            colorKey3_color = r.readT(4)
            colorKey1_colorAlpha = r.readLESingle()
            colorKey2_colorAlpha = r.readLESingle()
            colorKey3_colorAlpha = r.readLESingle()
            colorKey1_colorKeyTime = r.readLESingle()
            colorKey2_colorKeyTime = r.readLESingle()
            colorKey3_colorKeyTime = r.readLESingle()
        }
    }

    public override var description: String { return "EFSH: \(EDID)" }
    public var EDID: STRVField = STRVField_empty // Editor ID
    public var ICON: FILEField! // Fill Texture
    public var ICO2: FILEField! // Particle Shader Texture
    public var DATA: DATAField! // Data
    
    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        switch type {
        case "EDID": EDID = r.readSTRV(dataSize)
        case "ICON": ICON = r.readSTRV(dataSize)
        case "ICO2": ICO2 = r.readSTRV(dataSize)
        case "DATA": DATA = DATAField(r, dataSize)
        default: return false
        }
        return true
    }
}
